nodeJS/express/using_postgresql.md

```diff
@@ -0,0 +1,309 @@
+
+Invoke the above two functions in the specific controllers (you might have different function names etc. The important thing is to understand how the db functions are invoked):
+
+```javascript
+const db = require("../db/queries");
+
+async function getUsernames(req, res) {
+  const usernames = await db.getAllUsernames();
+  console.log("Usernames: ", usernames);
+  res.send("Usernames: " + usernames.map(user => user.username).join(", "));
+}
+
+async function createUsernameGet(req, res) {
+  // render the form
+}
+
+async function createUsernamePost(req, res) {
+  const { username } = req.body;
+  await db.insertUsername(username);
+  res.redirect("/");
+}
+
+module.exports = {
+  getUsernames,
+  createUsernameGet,
+  createUsernamePost
+};
+```
+
+Take your app for a spin, hopefully it works as expected.
+
+### Populate the db via a script
+
+You might have noticed how cumbersome it is to create a table and populate it with data. Luckily, we have the power of c(n)ode by our side, let's automate it via a script. Create a new file `db/populatedb.js`.
+
+```javascript
+#! /usr/bin/env node
+
+const { Client } = require("pg");
+
+const SQL = `
+CREATE TABLE IF NOT EXISTS usernames (
+  id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
+  username VARCHAR ( 255 )
+);
+
+INSERT INTO usernames (username) 
+VALUES
+  ('Bryan'),
+  ('Odin'),
+  ('Damon');
+`;
+
+async function main() {
+  console.log("seeding...");
+  const client = new Client({
+    connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users",
+  });
+  await client.connect();
+  await client.query(SQL);
+  await client.end();
+  console.log("done");
+}
+
+main();
+```
+
+Login to the PostgreSQL shell, connect to the `top_users` db, and drop the `usernames` table:
+
+```sql
+DROP TABLE usernames;
+```
+
+You can then run this script via `node db/populatedb.js`, or add it as a [`script` in `package.json`](https://stackoverflow.com/a/36433748).
+
+Do note that the script is designed to be ran only once.
+
+#### Populating production dbs
+
+We've hardcoded our local db connection information in the script. Hence, the script only populates our local db. We need a way to populate our production db as well. One way to this is by using environment variables, though this leads to unnecessary hassle. Why? Because now, the script can only populate the production db on the production server i.e. we'll need to access the production server's cli to run the script. Or, we could sneakily edit our environment file to point to the production db and run the script on our machine, and revert it.
+
+We should aim to make our script as independent from our codebase as possible.
+
+A far more painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [`process.argv`](https://nodejs.org/docs/latest/api/process.html#processargv).
```

Comment:
Nit:
```suggestion
You can then run this script via `node db/populatedb.js`, or add it as a [script in package.json](https://stackoverflow.com/a/36433748).

Do note that the script is designed to be ran only once.

#### Populating production dbs

We've hardcoded our local db connection information in the script. Hence, the script only populates our local db. We need a way to populate our production db as well. One way to this is by using environment variables, though this leads to unnecessary hassle. Why? Because now, the script can only populate the production db on the production server i.e. we'll need to access the production server's cli to run the script. Or, we could sneakily edit our environment file to point to the production db and run the script on our machine, and revert it.

We should aim to make our script as independent from our codebase as possible.

A far more painless approach is providing the connection information as an argument to the script. This way, we can run the script for local db as well as production db on our machine. You can access arguments via [process.argv](https://nodejs.org/docs/latest/api/process.html#processargv).
```
Do you think it's better to not use backticks for these two?

![image](https://github.com/TheOdinProject/curriculum/assets/75290989/0a5d8adf-9eca-471d-aa0b-a159f87f5ac3)

![image](https://github.com/TheOdinProject/curriculum/assets/75290989/d324079b-cde1-4e58-b961-084e9e021fcc)